### PR TITLE
Bump Python version in complex kernel to 3.12

### PR DESCRIPTION
**Pull request by jorgensd**
Since we are using ubuntu:24.04 as base now Python version is 3.12

### CODE REVIEW

File: docker/Dockerfile.end-user [hunk 1]
```diff
@@ -89,8 +89,8 @@ ONBUILD RUN cd dolfinx && \
     ninja install && \
     cd ../python && \
     PETSC_ARCH=linux-gnu-real64-32 pip install --break-system-packages -v \
-      --config-settings=cmake.build-type="${DOLFINX_CMAKE_BUILD_TYPE}" --config-settings=install.strip=false --no-build-isolation --check-build-dependencies \
-      --target /usr/local/dolfinx-real/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir . && \
+    --config-settings=cmake.build-type="${DOLFINX_CMAKE_BUILD_TYPE}" --config-settings=install.strip=false --no-build-isolation --check-build-dependencies \
+    --target /usr/local/dolfinx-real/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir . && \
```
Comment (jorgensd) [inline docker/Dockerfile.end-user:93]:
```suggestion
      --config-settings=cmake.build-type="${DOLFINX_CMAKE_BUILD_TYPE}" --config-settings=install.strip=false --no-build-isolation --check-build-dependencies \
      --target /usr/local/dolfinx-real/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir . && \
```

File: docker/Dockerfile.end-user [hunk 2]
```diff
@@ -100,8 +100,8 @@ ONBUILD RUN cd dolfinx && \
     . /usr/local/dolfinx-complex/lib/dolfinx/dolfinx.conf && \
     cd ../python && \
     PETSC_ARCH=linux-gnu-complex128-32 pip install --break-system-packages -v \
-      --config-settings=cmake.build-type="${DOLFINX_CMAKE_BUILD_TYPE}" --config-settings=install.strip=false --no-build-isolation --check-build-dependencies \
-      --target /usr/local/dolfinx-complex/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir .
+    --config-settings=cmake.build-type="${DOLFINX_CMAKE_BUILD_TYPE}" --config-settings=install.strip=false --no-build-isolation --check-build-dependencies \
+    --target /usr/local/dolfinx-complex/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir .
```
Comment (jorgensd) [inline docker/Dockerfile.end-user:104]:
```suggestion
      --config-settings=cmake.build-type="${DOLFINX_CMAKE_BUILD_TYPE}" --config-settings=install.strip=false --no-build-isolation --check-build-dependencies \
      --target /usr/local/dolfinx-complex/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir .
```

**Comment by garth-wells** [comment]
Is there something cleverer we could do so we don't have to change this for each Python update? I had no idea that a change was needed in `docker/complex-kernel.json`.

**Review by garth-wells — APPROVED** [review]

**Comment by jorgensd** [comment]
> Is there something cleverer we could do so we don't have to change this for each Python update? I had no idea that a change was needed in `docker/complex-kernel.json`.

We tried to reduce the number of places we have this in the code, but last time around I couldn't find a better solution.
(i fixed in the kernels in: https://github.com/FEniCS/dolfinx/pull/2211)

**Pull request merged**